File: backend/app/models/prediction.py
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, Integer, JSON, String, func
from sqlalchemy.orm import Mapped, mapped_column

from ..db.base import Base


class PredictionResult(Base):
    __tablename__ = "prediction_result"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    device_id: Mapped[str] = mapped_column(ForeignKey("device_info.device_id", ondelete="CASCADE", onupdate="CASCADE"), index=True)
    predict_time: Mapped[datetime] = mapped_column(DateTime, index=True)
    target_timestamp: Mapped[datetime | None] = mapped_column(DateTime)
    forecast_horizon: Mapped[int | None] = mapped_column(Integer)
    fault_type: Mapped[str | None] = mapped_column(String(32), index=True)
    probability: Mapped[float | None] = mapped_column(Float)
    p_no_failure: Mapped[float | None] = mapped_column(Float)
    p_heat: Mapped[float | None] = mapped_column(Float)
    p_power: Mapped[float | None] = mapped_column(Float)
    p_overstrain: Mapped[float | None] = mapped_column(Float)
    p_tool_wear: Mapped[float | None] = mapped_column(Float)
    predicted_params: Mapped[dict | None] = mapped_column(JSON)
    model_version: Mapped[str | None] = mapped_column(String(32))
    create_time: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

File: frontend/src/types/prediction.ts
import type { HealthEvaluation } from "./health";

export interface PredictionResult {
  id: number;
  device_id: string;
  predict_time: string;
  target_timestamp: string | null;
  forecast_horizon: number | null;
  fault_type: string | null;
  probability: number | null;
  p_no_failure: number | null;
  p_heat: number | null;
  p_power: number | null;
  p_overstrain: number | null;
  p_tool_wear: number | null;
  predicted_params: Record<string, unknown> | null;
  model_version: string | null;
}

export interface FaultProbability {
  device_id: string;
  p_no_failure: number;
  p_heat: number;
  p_power: number;
  p_overstrain: number;
  p_tool_wear: number;
}

export interface PredictionRun {
  prediction: PredictionResult;
  health: HealthEvaluation;
  stage1_query_result: Record<string, unknown>;
}

export interface PredictionTriggerRequest {
  query_wear?: number;
  machine_type?: "L" | "M" | "H" | "l" | "m" | "h";
  history_csv?: string;
}
